### PR TITLE
Added Ruler support for RulesAPI; Refactored Manager.

### PR DESCRIPTION
**Pull request by bwplotka**
As per: https://thanos.io/proposals/202003_thanos_rules_federation.md/

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

cc @s-urbaniak

### CODE REVIEW

File: cmd/thanos/rule.go [hunk 1]
```diff
@@ -386,69 +386,52 @@ func runRule(
...
-		}
+		// Run rule evaluation and alert notifications.
 		st := tsdb.Adapter(db, 0)
+		logger = log.With(logger, "component", "rules")
+		ruleMgr = thanosrules.NewManager(
+			tracing.ContextWithTracer(context.Background(), tracer),
+			reg,
+			dataDir,
+			rules.ManagerOptions{
+				NotifyFunc: func(ctx context.Context, expr string, alerts ...*rules.Alert) {
```
Comment (s-urbaniak) [inline cmd/thanos/rule.go:402]: nit: this is a long function, can we un-inline it for readability?

File: cmd/thanos/rule.go [hunk 2]
```diff
@@ -687,57 +670,59 @@ func removeDuplicateQueryEndpoints(logger log.Logger, duplicatedQueriers prometh
...
+					})
+					span.Finish()
+
+					if err != nil {
+						level.Error(logger).Log("err", err, "query", q)
+						continue
+					}
+					if len(warns) > 0 {
+						ruleEvalWarnings.WithLabelValues(strings.ToLower(partialResponseStrategy.String())).Inc()
+						// TODO(bwplotka): Propagate those to UI, probably requires changing rule manager code ):
```
Comment (s-urbaniak) [inline cmd/thanos/rule.go:718]: Is this something we should track as a dedicated issue?

**Comment by s-urbaniak** [comment]
looking great so far!

File: cmd/thanos/rule.go (hunk 2 above)
Comment (bwplotka) [reply cmd/thanos/rule.go:718]: I miss this, let's see

Comment (bwplotka) [reply cmd/thanos/rule.go:731]: I think it's fine for now ,not a biggy

File: pkg/rules/manager.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	for _, g := range m.RuleGroups() {
+		res = append(res, g.toProto())
+	}
+	return res
+}
+
+func (m *Manager) RuleGroups() []Group {
+	m.mtx.RLock()
+	defer m.mtx.RUnlock()
+	var res []Group
```
Comment (povilasv) [inline pkg/rules/manager.go:167]: consider preallocating slice

File: pkg/rules/manager.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+func (m *Manager) Stop() {
+	for _, mgr := range m.mgrs {
+		mgr.Stop()
+	}
+}
+
+func (m *Manager) protoRuleGroups() []*rulespb.RuleGroup {
+	var res []*rulespb.RuleGroup
```
Comment (povilasv) [inline pkg/rules/manager.go:157]: consider preallocating slice

File: pkg/rules/manager.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+				originalFile:            m.ruleFiles[group.File()],
+				PartialResponseStrategy: s,
+			})
+		}
+	}
+	return res
+}
+
+func (m *Manager) Active() []*rulespb.AlertInstance {
+	var res []*rulespb.AlertInstance
```
Comment (povilasv) [inline pkg/rules/manager.go:181]: ditto

File: pkg/rules/manager.go (hunk 3 above)
Comment (bwplotka) [reply pkg/rules/manager.go:168]: I cannot, I don't know number that managers have.

File: pkg/rules/manager.go (hunk 5 above)
Comment (bwplotka) [reply pkg/rules/manager.go:182]: Ditto, we don't have info upfront.

**Comment by s-urbaniak** [comment]
@bwplotka ping :-) can we merge once tests are fixed, i'd love to rebase on these changes :green_heart:

**Comment by bwplotka** [comment]
@s-urbaniak It's not as easy as it looks. I changed Prometheus version, so all deps will be reinstalled on CI causing major slowdown. We need to update CI but in mean time Go version changed on master.. :facepalm: 

Plus I am fixing other things, Now. I would propose following:

Since we have approval on design, let's merge CURRENT feature branch to master (or rather. propose a PR first). Let's rebase / pull latest master as well. Can we do it? :hugs: With this it would be easier for me to rebase my PRs on this (since this is actually chained with https://github.com/thanos-io/thanos/pull/2558 (!), then fix  tests, once we have Go version. Ideally we could merge feature branch now and iterate on master from now! (:

**Comment by s-urbaniak** [comment]
That would be a question for the remaining maintainers I guess, fine from my side, but I don't know what the policy is regarding master feature consistency.

**Comment by bwplotka** [comment]
So far things can be in progress. E.g see the receiver work. It was done on master.

**Comment by bwplotka** [comment]
As long as our new flags are hidden

**Comment by bwplotka** [comment]
But yea, let's just rebase/ pull master to feature branch anyway =D

**Comment by brancz** [comment]
I'm fine with merging things in steps as long as the flags are hidden until the feature is done and out of experimental.

**Comment by bwplotka** [comment]
Should be good to go @s-urbaniak

**Comment by s-urbaniak** [comment]
lgtm modulo the tests! :green_heart:

**Comment by bwplotka** [comment]
Looks like there is some deduplication bug, but you totally rewrote this in next PR: 

So I am skipping TestRulesAPI_Fanout test here. Let's enable on feature branch.

**Pull request merged**